### PR TITLE
[ACS] `az aks create/update`: Introduce changes for Azure container storage in ACS cli

### PR DESCRIPTION
**Pull request by mukhoakash**
**Related command**

**Description**

**Testing Guide**

**History Notes**

[Component Name 1] BREAKING CHANGE: `az command a`: Make some customer-facing breaking change
[Component Name 2] `az command b`: Add some customer-facing feature

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
azure-container-storage

**Comment by FumingZhang** [comment]
Could you please fix the failed CI checks?

File: src/azure-cli/azure/cli/command_modules/acs/azurecontainerstorage/acstor_ops.py [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+        except Exception as ex:  # pylint: disable=broad-except
+            config_settings = [
+                {"cli.storagePool.disable.validation": False},
+                {"cli.storagePool.disable.type": ""},
+            ]
+
+            err_msg = "Validation failed. " \
+                    "Please ensure that storagepools are not being used. " \
+                    "Unable to perform disable Azure Container Storage operation. " \
+                    "Reseting cluster state."
```
Comment (zioproto) [inline src/azure-cli/azure/cli/command_modules/acs/azurecontainerstorage/acstor_ops.py:269]: nit: "Resetting"

File: src/azure-cli/azure/cli/command_modules/acs/azurecontainerstorage/acstor_ops.py [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+            # Also, unset the type of storagepool which was supposed to disabled.
+            if storage_pool_type == CONST_STORAGE_POOL_TYPE_AZURE_DISK:
+                update_settings.append({"cli.storagePool.azureDisk.enabled": True})
+            elif storage_pool_type == CONST_STORAGE_POOL_TYPE_ELASTIC_SAN:
+                update_settings.append({"cli.storagePool.elasticSan.enabled": True})
+            elif storage_pool_type == CONST_STORAGE_POOL_TYPE_EPHEMERAL_DISK:
+                update_settings.append({"cli.storagePool.ephemeralDisk.enabled": True})
+            disable_op_failure = True
+
+        # Since we are just reseting the cluster state,
```
Comment (zioproto) [inline src/azure-cli/azure/cli/command_modules/acs/azurecontainerstorage/acstor_ops.py:401]: nit: resetting

File: src/azure-cli/azure/cli/command_modules/acs/azurecontainerstorage/acstor_ops.py [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+            LongRunningOperation(cmd.cli_ctx)(disable_op_result)
+        except Exception as disable_ex:
+            logger.error(
+                "Failure observed while disabling Azure Container Storage storagepool type: %s.\nError: %s"
+                "Reseting cluster state.", storage_pool_type, disable_ex
+            ) from disable_ex
+
+            # If disabling type of storagepool in Azure Container Storage failed,
+            # define the existing resource values for ioEngine and hugepages for
+            # reseting the cluster state.
```
Comment (zioproto) [inline src/azure-cli/azure/cli/command_modules/acs/azurecontainerstorage/acstor_ops.py:382]: nit: resetting

File: src/azure-cli/azure/cli/command_modules/acs/azurecontainerstorage/_validators.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+
+# _validate_nodepool_names validates that the nodepool_list is a comma separated
+# string consisting of valid nodepool names i.e. a lower alphanumeric
+# characters and the first character should be lowercase letter.
+def _validate_nodepool_names(nodepool_names, agentpool_details):
+    pattern = r'^[a-z][a-z0-9]*(?:,[a-z][a-z0-9]*)*$'
+    if re.fullmatch(pattern, nodepool_names) is None:
+        raise InvalidArgumentValueError(
+            "Invalid --azure-container-storage-nodepools value. "
+            "Accepted value is a comma separated string of valid nodepool "
```
Comment (zioproto) [inline src/azure-cli/azure/cli/command_modules/acs/azurecontainerstorage/_validators.py:224]: consistency: "node pools" 
should be two distinct words in English, consistent with the AKS docs:
https://learn.microsoft.com/en-us/azure/aks/create-node-pools

in the command it is correct to use the form `nodepool`

File: src/azure-cli/azure/cli/command_modules/acs/azurecontainerstorage/_validators.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+
+    nodepool_list = nodepool_names.split(',')
+    for nodepool in nodepool_list:
+        if nodepool not in agentpool_names:
+            if len(agentpool_names) > 1:
+                raise InvalidArgumentValueError(
+                    f'Nodepool: {nodepool} not found. '
+                    'Please provide a comma separated string of existing nodepool names '
+                    'in --azure-container-storage-nodepools.'
+                    f"\nNodepools available in the cluster are: {', '.join(agentpool_names)}."
```
Comment (zioproto) [inline src/azure-cli/azure/cli/command_modules/acs/azurecontainerstorage/_validators.py:241]: consistency: Node pools

File: src/azure-cli/azure/cli/command_modules/acs/azurecontainerstorage/_validators.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+
+def _validate_nodepools_for_ephemeral_nvme_storagepool(agentpool_details):
+    for agentpool in agentpool_details:
+        vm_size = agentpool.get("vm_size")
+        if vm_size is not None and vm_size.lower().startswith('standard_l'):
+            break
+    else:
+        raise ArgumentUsageError(
+            f'Cannot set --storage-pool-option as {CONST_STORAGE_POOL_OPTION_NVME} '
+            'as no supporting nodepool found which can support ephemeral NVMe disk.'
```
Comment (zioproto) [inline src/azure-cli/azure/cli/command_modules/acs/azurecontainerstorage/_validators.py:274]: readibility: "as none of the node pools can support ephemeral NVMe disk."

File: src/azure-cli/azure/cli/command_modules/acs/azurecontainerstorage/acstor_ops.py [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+            err_msg = "Validation failed. " \
+                    "Please ensure that storagepools are not being used. " \
+                    "Unable to perform disable Azure Container Storage operation. " \
+                    "Reseting cluster state."
+
+            if storage_pool_type != CONST_STORAGE_POOL_TYPE_ALL:
+                err_msg = "Validation failed. " \
+                        f"Please ensure that storagepools of type {storage_pool_type} are not being used. " \
+                        f"Unable to perform disable Azure Container Storage operation. " \
+                        "Reseting cluster state."
```
Comment (zioproto) [inline src/azure-cli/azure/cli/command_modules/acs/azurecontainerstorage/acstor_ops.py:275]: nit: Resetting

File: src/azure-cli/azure/cli/command_modules/acs/azurecontainerstorage/acstor_ops.py [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+                "managedClusters",
+                configuration_settings=config_settings,
+                yes=True,
+                no_wait=True,
+            )
+
+            if "pre-upgrade hooks failed" in str(ex):
+                raise UnknownError(err_msg) from ex
+            raise UnknownError(
+                "Validation failed. Unable to perform Azure Container Storage operation. Reseting cluster state."
```
Comment (zioproto) [inline src/azure-cli/azure/cli/command_modules/acs/azurecontainerstorage/acstor_ops.py:292]: nit: Resetting

File: src/azure-cli/azure/cli/command_modules/acs/azurecontainerstorage/acstor_ops.py [hunk 9]
```diff
@@ -0,0 +1,416 @@
...
+                "managedClusters",
+                configuration_settings=config_settings,
+                yes=True,
+                no_wait=False,
+            )
+            LongRunningOperation(cmd.cli_ctx)(disable_op_result)
+        except Exception as disable_ex:
+            logger.error(
+                "Failure observed while disabling Azure Container Storage storagepool type: %s.\nError: %s"
+                "Reseting cluster state.", storage_pool_type, disable_ex
```
Comment (zioproto) [inline src/azure-cli/azure/cli/command_modules/acs/azurecontainerstorage/acstor_ops.py:377]: nit: Resetting

**Review by andyliuliming — CHANGES_REQUESTED** [review]
not a aks addon. should not be added into the az aks cmd

**Review by andyliuliming — COMMENTED** [review]
reset the request changes.

**Comment by yanzhudd** [comment]
Please fix the CI issues

**Comment by mukhoakash** [comment]
> Please fix the CI issues

CIs fixed.

**Comment by mukhoakash** [comment]
[Live test ](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=89653635&view=results) triggered:
- test_aks_create_with_azurecontainerstorage
- test_aks_update_with_azurecontainerstorage

**Comment by yanzhudd** [comment]
Please note that CLI is only responsible for the code style and specification, but not business logic. So it it better to get approval from the service team first. 
Besides, there appears to be one pending change request.

**Comment by mukhoakash** [comment]
> [Live test ](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=89653635&view=results) triggered:
> 
> * test_aks_create_with_azurecontainerstorage
> * test_aks_update_with_azurecontainerstorage

Link to updated [live tests](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=89736569&view=results).

File: src/azure-cli/azure/cli/command_modules/acs/managed_cluster_decorator.py [hunk 10]
```diff
@@ -8053,6 +8375,82 @@ def postprocessing_after_mc_created(self, cluster: ManagedCluster) -> None:
...
+        disable_azure_container_storage = self.context.get_intermediate("disable_azure_container_storage")
+        is_extension_installed = self.context.get_intermediate("is_extension_installed")
+        is_azureDisk_enabled = self.context.get_intermediate("is_azureDisk_enabled")
+        is_elasticSan_enabled = self.context.get_intermediate("is_elasticSan_enabled")
+        is_ephemeralDisk_localssd_enabled = self.context.get_intermediate("is_ephemeralDisk_localssd_enabled")
+        is_ephemeralDisk_nvme_enabled = self.context.get_intermediate("is_ephemeralDisk_nvme_enabled")
+        current_core_value = self.context.get_intermediate("current_core_value")
+        pool_option = self.context.raw_param.get("storage_pool_option")
+
+        print("HERE?")
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/managed_cluster_decorator.py:8388]: remove this?

File: src/azure-cli/azure/cli/command_modules/acs/managed_cluster_decorator.py [hunk 11]
```diff
@@ -8053,6 +8375,82 @@ def postprocessing_after_mc_created(self, cluster: ManagedCluster) -> None:
...
+
+        # enable azure container storage
+        if enable_azure_container_storage:
+            print("HERE GAIN?")
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/managed_cluster_decorator.py:8392]: and other prints?

**Review by FumingZhang — APPROVED** [review]
lgtm

please remove the prints used for debug purpose

**Review by andyliuliming — APPROVED** [review]

**Review by yanzhudd — APPROVED** [review]

**Pull request merged**